File: skills/veracity-auth-net/assets/SafeRedirect.cs
// Template: Open-redirect guard (CWE-601) for the Veracity OIDC sign-in returnUrl.
// File: Extensions/SafeRedirect.cs
//
// The `returnUrl` accepted by /auth/challenge is caller-supplied and must never drive a
// redirect without validation. `Uri.IsWellFormedUriString(returnUrl, UriKind.Relative)`
// is NOT sufficient on its own — a protocol-relative value such as "//evil.com" is a
// well-formed relative URI, yet the browser treats it as an absolute cross-origin URL.
// This mirrors ASP.NET Core's built-in Url.IsLocalUrl logic: only single-slash
// root-relative paths (and "~/" app-relative paths) are accepted; anything absolute,
// protocol-relative ("//host"), or backslash-obfuscated ("/\\host") collapses to "/".

internal static class SafeRedirect
{
    /// <summary>Returns <paramref name="returnUrl"/> when it is a safe local path, otherwise <paramref name="fallback"/>.</summary>
    public static string SanitizeReturnUrl(string? returnUrl, string fallback = "/")
        => IsLocalUrl(returnUrl) ? returnUrl! : fallback;

    /// <summary>Local-URL check equivalent to ASP.NET Core's Url.IsLocalUrl.</summary>
    public static bool IsLocalUrl(string? url)
    {
        if (string.IsNullOrEmpty(url))
            return false;

        // Allows "/" and "/foo" but not "//" or "/\" (protocol-relative / backslash tricks).
        if (url[0] == '/')
            return url.Length == 1 || (url[1] != '/' && url[1] != '\\');

        // Allows "~/" app-relative paths but not "~//" or "~/\".
        if (url[0] == '~' && url.Length > 1 && url[1] == '/')
            return url.Length == 2 || (url[2] != '/' && url[2] != '\\');

        return false;
    }
}

File: skills/veracity-auth-node/assets/nestjs/oidc/safeRedirect.ts
// Template (NestJS): src/auth/safeRedirect.ts
// Open-redirect guard (CWE-601). The OIDC `returnUrl` is caller-supplied, so it must
// never be used to build a redirect without validation. `safeReturnUrl` returns a
// guaranteed application-relative path: anything that is absolute, protocol-relative
// (`//host`), backslash-obfuscated, scheme-bearing, or otherwise not a plain
// root-relative path collapses to the safe `fallback` ("/").

export function safeReturnUrl(raw: unknown, fallback = "/"): string {
  if (typeof raw !== "string") return fallback;
  const value = raw.trim();
  // Must be a single-slash root-relative path.
  if (!value.startsWith("/")) return fallback;
  // Block protocol-relative ("//host") and backslash-obfuscated ("/\\host") forms.
  if (value.startsWith("//") || value.startsWith("/\\")) return fallback;
  // Reject backslashes, control characters, and any embedded scheme/host.
  if (value.includes("\\")) return fallback;
  if (/[\u0000-\u001f\u007f]/.test(value)) return fallback;
  if (value.includes("://")) return fallback;
  return value;
}
